lectures/pandas.md

```diff
@@ -205,6 +200,223 @@ To select rows and columns using a mixture of integers and labels, the `loc` att
 df.loc[df.index[2:5], ['country', 'tcgdp']]
 ```
 
+### Select Data by Conditions
+
+Instead of indexing rows and columns using integers, we can also obtain a sub-dataframe of our interests that satisfies certain (potentially complicated) conditions specified by us.
+
+This section demonstrates various ways to do that.
+
+The most straigntforward way is with the `[]` operator.
+
+```{code-cell} python3
+df[df.POP >= 20000]
+```
+
+To understand what is going on here, notice that `df.POP >= 20000` returns a series of boolean values.
+
+```{code-cell} python3
+df.POP >= 20000
+```
+
+In this case, `df[___]` takes a series of boolean values and only returns rows that correspond to the `True` values.
+
+Take one more example,
+
+```{code-cell} python3
+df[(df.country.isin(['Argentina', 'India', 'South Africa'])) & (df.POP > 40000)]
+```
+
+However, there is another way of doing the same thing, which can be slightly faster for large dataframes, with more natural syntax.
+
+```{code-cell} python3
+# the above is equivalent to 
+df.query("POP >= 20000")
+```
+
+```{code-cell} python3
+df.query("country in ['Argentina', 'India', 'South Africa'] and POP > 40000")
+```
+
+We can also allow arithmetic operations between different columns.
+
+```{code-cell} python3
+df[(df.cc + df.cg >= 80) & (df.POP <= 20000)]
+```
+
+```{code-cell} python3
+# the above is equivalent to 
+df.query("cc + cg >= 80 & POP <= 20000")
+```
+
+For example, we can use the conditioning to select the country with the largest household consumption - gdp share `cc`.
+
+```{code-cell} python3
+df.loc[df.cc == max(df.cc)]
+```
+
+When we only want to look at certain columns of a selected sub-dataframe, we can use the above conditions with the `.loc[__ , __]` command.
+
+The first argument takes the condition, while the second argument takes a list of columns we want to return.
+
+```{code-cell} python3
+df.loc[(df.cc + df.cg >= 80) & (df.POP <= 20000), ['country', 'year', 'POP']]
+```
+
+
+**Application: Subsetting Dataframe**
+
+Real world datasets can be [enoumous](https://developers.google.com/machine-learning/data-prep/construct/collect/data-size-quality).
+
+It is sometimes desirable to work with a subset of data for computational efficiency and reduce redundancy to save space.
```

Comment:
The grammar in this sentence is not quite right.  Maybe "...to enhance computational efficiency, reduce redundancy and save space"?